### ISSUE TITLE
Inconsistencies in date/time functions.

### ISSUE THREAD

**Issue by yoavnir** [yoavnir/specs2016#119]
d2tf() accepts time in microsconds but the manpage says it gets time in seconds.

**Comment by Gawesomer** [comment]
Hi there. I'm looking to make my first open source contribution.
May I work on this issue?
Just to make sure I understand the issue; you want seconds to be changed to microseconds in the manpage as well as any comments and docs in which d2tf is referenced. Is that right?
Should tf2d specifications be changed to reflect this modification or is its documentation correct?

**Comment by yoavnir** [comment]
A while back I converted the internal time format from floating-point seconds to integer microseconds. The reason was that I was getting different results for the same formatted time on Windows vs Linux -- not much, just a microsecond, but enough to make unit tests fail.

Anyway, the d2tf/tf2d functions should have consistent documentation. So the documentation should be fixed.

Even better would be to also add some seconds-based functions that get and return float -- perhaps tf2s / s2tf.

But sure, go ahead and fix this.  Thanks.  Please do it on the v0.6pre branch and send a PR

**Comment by Gawesomer** [comment]
Since the internal time is now being converted to microseconds, is %xf still used in the format for d2tf() to specify fractional seconds? Should it be kept in the documentation?

**Comment by yoavnir** [comment]
Sure. The internal time is expressed in microseconds so that the formatted time can have up to 6 decimal places after the number of seconds.

**Comment by yoavnir** [comment]
```
commit 932016fcc10acdd51885ee5f0a99e413164e2099
Author: Gawesomer <jj.jouband@gmail.com>
Date:   Fri Dec 6 11:28:05 2019 -0600

    Edited manpage and docs to reflect that tf2d() and d2tf() now deal with microseconds instead of seconds
```

**Comment by yoavnir** [comment]
The above commit fixes the manpage and .md docs.

@Gawesomer : Are you going to also add the tf2s and s2tf functions?  Either way, I'll leave the issue open for now.

**Comment by Gawesomer** [comment]
Sure, I'll give it a shot.

**Comment by Gawesomer** [comment]
In order to implement the testing for these new functions I looked back at the testing done for tf2d() and d2tf() and am a little confused by it. 
In ProcessingTest.cc a bunch of tests are run for d2tf() giving the input in seconds of '1545407296.548900' and expecting the output of 'Fri Dec 21 18:48:16 2018'. Since d2tf() now takes microseconds as input, this should be incorrect and I would have expected these tests to fail, yet when running the test suite it appears that all the tests have passed. 

Is this an issue or am I simply misunderstanding the testing?
It looks to me as though the VERIFY macro is hard coded with "The quick brown fox jumped over the   lazy dog".
How should I implement the testing for tf2s() and s2tf()?

**Comment by yoavnir** [comment]
You're confusing conversions with functions.

In ProcessingTest.cc there are tests for data fields, so you might have something like 
`w1 d2tf %S 1` - take the first word, convert it from decimal to a format and print at offset 1.

Functions are tested in ALUUnitTest.cc, and you have something like
`print "d2tf(word(1),'%S')" 1` - which does the exact same thing.

So tests for functions should be in ALUUnitTest.cc.

That said, I agree it's confusing that a conversion and function with the same name will use different units. How do you think this should be reconciled?

**Comment by Gawesomer** [comment]
To remedy this, as well as adding s2tf() and tf2s() functions we might also change the d2tf and tf2d conversions to take microseconds as input as well as implement tf2s and s2tf conversions which take seconds as input.
This would offer two conversion methods with inputs matching that of their respective functions.
Then again, I'm not sure that two conversion methods are necessary.

**Comment by yoavnir** [comment]
OK. Let's see. The microseconds-since-the-epoch value is good as an internal format, but next to useless for output. Seconds-since-the-epoch is also not very user friendly, but it is a common value used in many file formats.

So the d2tf and tf2d conversions are only useful in reading some timestamp into the internal format, usually into a field identifier or from an expression.  If we add s2tf and tf2s conversions, they will not be useful for the internal format. I don't think they'll be very useful for output, because file formats that use seconds-since-the-epoch tend to express that number as a binary, not decimal representation, so I don't think they'd be very useful.

Where seconds-since-the-epoch *is* useful is for doing calculations and time differences, like taking the timestamp from two successive records and calculating how many seconds passed between them. It can all be done with functions. The question is which is more natural:
```
w1-2 tf2s @myformat a:
if "!first()" then
    /diff:/ 1 print "a-#0" nw /seconds/ nw
endif
set "#0:=a"
```
vs
```
set "#1:=tf2s(words(1,2),@myformat)"
if "!first()" then
    /diff:/ 1 print "#1-#0" nw /seconds/ nw
endif
set "#0:=a"
```
Or even
```
w1-2 a:
set "#1:=tf2s(a,@myformat)"
if "!first()" then
    /diff:/ 1 print "#1-#0" nw /seconds/ nw
endif
set "#0:=a"
```
I'm sort of leaning towards the latter ones, so I think functions are the way to go, and we can leave the conversions as they are. 

But I'll leave it to you.

**Comment by Gawesomer** [comment]
I've implemented tf2s() and s2tf() as ALU functions.
Hope that works.

We might want to change the name of d2tf() and tf2d() ALU functions to something else, so that they are no longer confused with the conversions. Maybe mcs2tf() and tf2mcs() (where mcs stands for microseconds) might do the trick. 
What do you think?

**Comment by yoavnir** [comment]
Seems to work, but it's still inconsistent that tf2d/d2tf have different meanings in conversions and ALU:

```
% specs "Tue Dec 10 22:08:59 2019" tf2d %c 1
1576008539.000000
```

but
```
 % specs print "tf2d('Tue Dec 10 22:08:59 2019', '%c')" 1
1576008539000000
```

while
```
% specs print "tf2s('Tue Dec 10 22:08:59 2019', '%c')" 1
1576008539
```

So yes, I like your idea of renaming the *functions* d2tf/tf2d so there are never two things that have the same name but do different things.  The only part I don't like is that the TOD clock does not fit the conversions:
```
% specs TOD 1
1576008377084795
```
That's a microseconds value and so cannot be converted with the tf2d/d2tf:
```
% specs TOD d2tf %c 1
Sun Dec 21 22:20:00 -290308
```

**Comment by Gawesomer** [comment]
Added mcs2tf() and tf2mcs() conversions in PR #122. 
I've also changed the names of the d2tf() and tf2d() conversions to be more explicit.
There are now mcs2tf() and tf2mcs() conversions and ALU functions which deal with microseconds, similarly for s2tf() and tf2s() which deal with seconds.
The documentation has been edited. I hope I didn't miss anything.
The TOD should now be convertable using the mcs2tf() conversion as such:

`$ specs TOD mcs2tf %c 1
Wed Dec 11 11:57:45 2019`
